### ISSUE TITLE
Think about how to express doubts in the result

### ISSUE THREAD

**Issue by Martoon-00** [serokell/tzbot#36]
# Clarification and motivation

Imagine a scenario: today (at Jan 16) I write a message that mentions 12:00 of Nov 23. And we should try to guess an year.
We already have a guessing logic after #2, and the question is: how should we express when we are not sure about the result in such cases?

And there are options:
1. Never doubt, just post the guess.
2. Add a note that we are unsure.
3. Do not send any messages to other users, say the message author's that his timestamp is ambiguous and should be fixed.
4. Select between (2) and (3) depending on how sure we are in the result.

# Acceptance criteria

* We discussed the mentioned point and applied a solution.

**Comment by Martoon-00** [comment]
I think we have to keep the following facts in mind:
* Preferably, the Bot should guess the message author's intentions.
* If Bot mis-guesses and submits wrong dates (applying wrong timezone offsets), that would be a fatal mistake.
* Not all people double-check which year-moths the bot auto-inferred and may miss the bad guess.
* Currently Bot does not send any messages to the message's author, so author also has no chance to double-check the Bot's guess unless we apply special actions here.

That all suggests that (1) is too unsafe, and (2) should be thought over thoroughly to be made safe.

**Comment by Martoon-00** [comment]
> Actually it [does](https://github.com/serokell/tzbot/blob/8f24fa2646bd121d2370a8c34ddf71fa545eb3b4/src/TzBot/ProcessEvent.hs#L86). So we can also include our doubts to this "callback".

Ah I see, that's important to account.

---

One thing to clarify:

> "10am" in Asia/Tashkent is, possibly, 10:00, Wednesday, January 18 in Asia/Tashkent

here both timestamps are in the author's timezone, or the 1st one is in the author's timezone and the 2nd one is in the receiver's timezone?

**Comment by Martoon-00** [comment]
> But for the third situation we need something more "scary"

I think what you described for this case is mostly the perfect solution, agreed.

**Comment by YuriRomanowski** [comment]
> here both timestamps are in the author's timezone, or the 1st one is in the author's timezone and the 2nd one is in the receiver's timezone?

Yes, during development process I have to send messages to myself :)

**Comment by Martoon-00** [comment]
>  Yes, during development process I have to send messages to myself :)

Aha, I see. That's important, because if we quote the original timestamp and then print a translation, then there is no easy way for the receivers to verify whether the Bot's guess about the author's meant date was correct. And like, it's not clear what next actions should they take:
* Believe the Bot? That's a risk.
* Don't believe the bot's guess? Then we could just send "Sorry, too hard" with the same effect :)
* Double-check with the author? Then we could not notify the users and just ask the author to fix the timestamp to simplify things for users.

> Maybe we should handle such users in the same way as users in other timezones.

I personally think we should really go with uniform behaviour for all receivers, and make it smarter later.

**Comment by YuriRomanowski** [comment]
Guys, in the [PR](https://github.com/serokell/tzbot/pull/48) related to this issue @dcastro proposed quite different way of solving this issue that has nearly nothing common with the approach implemented there. Here is the quote:

> I personally find this construction a bit confusing:
> 
> > "10am in Europe/Moscow: _02:00 in America/Havana_ (possibly, Monday, January 30)
> 
> It sounds like we inferred the date for the _target_ of the conversion (02:00 in Havana).
> 
> But that's not true, what we did was infer the date for the _source_ of the conversion ("10am in Europe/Moscow"), and _then_ applied the conversion as usual.
> 
> I'd like to suggest an alternative solution. What we want is to make explicit whatever information the bot inferred, right?
> 
> In #36, I see you've [listed](https://github.com/serokell/tzbot/issues/36#issuecomment-1385700913) these 3 bits of information the bot sometimes has to infer:
> 
> > * We are trying to guess the month given the day of month;
> > * We are trying to guess the year given the day of month and the month;
> > * The time reference doesn't contain date references at all, and we choose either today or tomorrow.
> 
> Two other things we infer that are missing from that list:
> 
>     * Weekdays:  when a user says "10am on Wednesday" we infer they mean the _next_ Wednesday, and figure out what date that is
> 
>     * Timezones: when user says "10am", we infer they mean to refer to the timezone they're currently in.
> 
> 
> My suggestion is to simply add all the inferred information to the left-hand block.
> 
> For example, here we inferred the date (using the day of the week given by the sender) and the timezone:
> 
> ```
> Sender: let's meet at 10am on Wednesday
> Ephemeral: 
>   "10am on Wednesday", 1 February 2023 in Europe/Moscow                   
>                                         02:00, Monday, January 30, 2023 in America/Havana
> ```
> 
> Here, say it's currently 11am in the sender's timezone, so when they said "10am" we inferred they meant "10am tomorrow". We also inferred the timezone.
> 
> ```
> Sender: let's meet at 10am
> Ephemeral: 
>   "10am", 1 February 2023 in Europe/Moscow                   
>                                         02:00, Monday, January 30, 2023 in America/Havana
> ```
> 
> Here the date was explicit, so all we had to do was infer the timezone:
> 
> ```
> Sender: let's meet at 10am on 1/2/2023
> Ephemeral:
>   "10am on 1/2/2023" in Europe/Moscow                   
>                                         02:00, Monday, January 30, 2023 in America/Havana
> ```
> 
> Basically, all the explicit information goes inside quotes, followed by the inferred information:
> 
> ```
> "10am", 1 February 2023 in Europe/Moscow            
> |____|    |___________________________|
>    |                 |
> explicit          inferred
> ```
> 
> What do you think?
> 
> We already kinda do this for the inferred timezone (e.g. `"10am" in Europe/Moscow ...`), I'm suggesting we do it for all inferred information as well.

AFAIU, here we totally refuse to express doubts explicitly (via "possibly" or something), rather show to user what parts of the date we inferred by appending this information to the original time reference and not enclosing it with quotes. So, we guarantee that if the date was guessed right, the right-hand side of the ephemeral is also right, and if not, the time on the right can be wrong (because of the time shifts), and user can trust it if he believes that around this date time doesn't depend on the date.

Also, as discussed with @Martoon-00, we should try to warn the user somehow if the timeshift is actually somewhere near, e.g. by adding a warning as a note, which can contain informations about timeshifts for source or target (or both) timezones. And I would like us to discuss the conditions under which we add such a note. We could discuss this aspect here or create a different issue for this, what do you think guys? As for me, it looks like a good separate issue.

Also, @Martoon-00 created an issue https://github.com/serokell/tzbot/issues/53 which is partially covered in the mentioned PR https://github.com/serokell/tzbot/pull/48, it seems that would be a good solution just to close that PR and instead add some other PRs: one related to this issue (implementing new @dcastro proposed solution), another possibly related to warning about time shifts, and the latter implementing https://github.com/serokell/tzbot/issues/53 (I hope we'll figure out optimal solution for it). Does it all make sense guys?
